src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperBlockRewriter.cs

```diff
@@ -32,6 +34,21 @@ internal static class TagHelperBlockRewriter
                 {
                     return TagMode.StartTagOnly;
                 }
+
+                // Directive attribute will tolerate forms that don't work for tag helpers. For instance:
+                //
+                // <input @onclick="..."> vs <input onclick="..." />
+                //
+                // We don't want this to become an error just because you added a directive attribute.
+                if (descriptor.IsAnyComponentDocumentTagHelper() && !descriptor.IsComponentOrChildContentTagHelper())
```

Comment:
Can you explain this if condition? Not sure how this being true means it contains a directive attribute.

Reply:
The things we use tag helpers for:
- ITaghelper
- view components
- Components
- Component Child COntent
- Directive Attributes

The last 3 are included by the first condition, and the first two of those are excluded by the second condition. We could add a dedicated extension for this, but it would have the same logic internally probably.